akka-docs/rst/scala/persistence-schema-evolution.rst

```diff
@@ -0,0 +1,471 @@
+  - Allow the system to continue operating without large scale migrations to be applied,
+  - Allow the system to read "old" events from the underlying storage, however present them in a "new" view to the application logic,
+  - Transparently promote events to the latest versions during recovery (or queries) such that the business logic need not consider multiple versions of events
+
+.. _Immutability Changes Everything: http://www.cidrdb.org/cidr2015/Papers/CIDR15_Paper16.pdf
+.. _Event Sourcing: http://martinfowler.com/eaaDev/EventSourcing.html
+
+Types of schema evolution
+-------------------------
+Before we explain the various techniques that can be used to safely evolve the schema of your persistent events
+over time, we first need to define what the actual problem is, and what the typical styles of changes are.
+
+Since events are never deleted, we need to have a way to be able to recover (read) old events, in such way
+that does not force the ``PersistentActor`` to be aware of all possible versions of an event that it may have
+persisted in the past. Instead, we want the Actors to work on some form of "latest" version of the event and provide some
+means of either converting old "versions" of stored events into this "latest" event type, or constantly evolve the event
+definition - in a backwards compatible way - such that the new deserialization code can still read old events.
+
+The most common schema changes you will likely are:
+
+- :ref:`adding a field to an event type <add-field-scala>`,
+- :ref:`remove or rename field in event type <rename-field-scala>`,
+- :ref:`remove event type <remove-event-class-scala>`,
+- :ref:`split event into multiple smaller events <split-large-event-into-smaller-scala>`.
+
+The following sections will explain some patterns which can be used to safely evolve your schema when facing those changes.
+
+Picking the right serialization format
+======================================
+
+Picking the serialization format is a very important decision you will have to make while building your application.
+It affects which kind of evolutions are simple (or hard) to do, how much work is required to add a new datatype, and,
+last but not least, how serialization performance serialization.
+
+If you find yourself realising you have picked "the wrong" serialization format, it is always possible to change
+the format used for storing new events, however you would have to keep the old deserialization code in order to
+be able to replay events that were persisted using the old serialization scheme. It is possible to "rebuild"
+an event-log from one serialization format to another one, however it may be a more involved process if you need
+to perform this on a live system.
+
+Binary serialization formats that we have seen work well for long-lived applications include the very flexible IDL based:
+`Google Protobuf`_, `Apache Thrift`_ or `Apache Avro`_ (for which schema evolution is more "entire schema" based, instead of
```

Comment:
instead of putting the Avro remark in () it might be easier to read if was a new sentence starting with "Avro schema evolution is..."